test/e2e/common/machinepoolnamelease.go

```diff
@@ -0,0 +1,78 @@
+package common
+
+import (
+	"context"
+	"fmt"
+	"github.com/openshift/hive/pkg/constants"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/watch"
+	clientwatch "k8s.io/client-go/tools/watch"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"time"
+
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1"
+
+	log "github.com/sirupsen/logrus"
+
+	"k8s.io/client-go/rest"
+	clientcache "k8s.io/client-go/tools/cache"
+)
+
+// WaitForMachinePoolNameLease waits for a lease to appear in the given namespace for the given pool.
+func WaitForMachinePoolNameLease(config *rest.Config, namespace, poolName string, timeOut time.Duration) (string, error) {
+	logger := log.WithField("pool", poolName)
+	logger.Infof("Waiting for MachinePoolNameLease")
+	scheme, err := hivev1.SchemeBuilder.Build()
+
+	waitContext, cancel := context.WithTimeout(context.Background(), timeOut)
+	defer cancel()
+
+	gvk, err := apiutil.GVKForObject(&hivev1.MachinePoolNameLease{}, scheme)
+	if err != nil {
+		return "", errors.Wrap(err, "could not get the GVK for clusterprovisions")
+	}
+	codecs := serializer.NewCodecFactory(scheme)
+	restClient, err := apiutil.RESTClientForGVK(gvk, config, codecs)
+	if err != nil {
+		return "", errors.Wrap(err, "could not create REST client")
+	}
+
+	optionsModifier := func(options *metav1.ListOptions) {
+		options.LabelSelector = fmt.Sprintf("%s=%s", constants.MachinePoolNameLabel, poolName)
+	}
+
+	var lease *hivev1.MachinePoolNameLease
+	var ok bool
+	_, err = clientwatch.UntilWithSync(
+		waitContext,
+		clientcache.NewFilteredListWatchFromClient(
+			restClient,
+			"machinepoolnameleases",
+			namespace,
+			optionsModifier,
+		),
+		&hivev1.MachinePoolNameLease{},
+		nil,
+		func(event watch.Event) (bool, error) {
+			switch event.Type {
+			case watch.Added, watch.Modified, watch.Deleted:
+				lease, ok = event.Object.(*hivev1.MachinePoolNameLease)
```

Comment:
I don't think that you need to do this part here. I would move it to after the `UnitlWithSync` call. If there is an object received in the event that is not a MachinePoolNameLease, then that is reason enough to fail the test.

```
event, err := clientwatch.UntilWithSync(
  ...
  func(event watch.Event) (bool, error) {
    return event.Type == watch.Added || event.Type == watch.Modified
  },
)
if err != nil {
  return "", errors.Wrap(err, "timeout waiting for MachinePoolNameLease")
}
lease, ok := event.Object.(*hivev1.MachinePoolNameLease)
if !ok {
  return "", errors.Fmt("expected a MachinePoolNameLease object but got a %T object instead", event.Object)
}
logger.Infof("found MachinePoolName lease: %s", lease.Name)
return lease.Name, nil
```

Reply:
Looks better.